File: dotnet/src/Bewerbung.Api/StellenangebotService.cs
namespace Bewerbung.Api;

/// <summary>
/// Use-Case-Schicht fuer Stellenangebote: reine Fachlogik, kennt
/// weder DB noch HTTP.
///
/// Wichtigste Geschaeftsregel:
///   Eine neue Stelle startet IMMER mit Status 'ENTWURF'.
///   Ein vom Aufrufer gelieferter Status wird bewusst ignoriert.
/// </summary>
public sealed class StellenangebotService
{
    public const string StatusEntwurf = "ENTWURF";

    public static readonly IReadOnlyList<string> Arten = new[]
    {
        "FESTANSTELLUNG", "AZUBI", "MINIJOB", "WERKSTUDENT", "PRAKTIKUM",
    };

    public static readonly IReadOnlyList<string> Stati = new[]
    {
        "ENTWURF", "VEROEFFENTLICHT", "GESCHLOSSEN", "ARCHIVIERT",
    };

    private readonly IStellenangebotRepository _repo;

    public StellenangebotService(IStellenangebotRepository repo)
    {
        _repo = repo;
    }

    public async Task<AnlegenResult> AnlegenAsync(StelleHttpInput input)
    {
        Validate(input);

        var titel = input.Titel!.Trim();
        var beschreibung = string.IsNullOrWhiteSpace(input.Beschreibung)
            ? null
            : input.Beschreibung!.Trim();
        var art = string.IsNullOrWhiteSpace(input.Art) ? "FESTANSTELLUNG" : input.Art!;

        // Geschaeftsregel: neue Stellen starten IMMER als ENTWURF.
        var status = StatusEntwurf;

        var id = await _repo.InsertStelleAsync(new StelleInput(
            titel, beschreibung, art, status));

        return new AnlegenResult(id, titel, art, status);
    }

    public Task<IReadOnlyList<IDictionary<string, object?>>> ListeAsync(string? status)
    {
        if (status is not null && !Stati.Contains(status))
        {
            throw new ValidationException(new[]
            {
                "Parameter 'status' ist kein gueltiger Stellenstatus.",
            });
        }
        return _repo.ListStellenAsync(status);
    }

    private static void Validate(StelleHttpInput i)
    {
        var errors = new List<string>();

        var titel = (i.Titel ?? string.Empty).Trim();
        if (string.IsNullOrEmpty(titel))
        {
            errors.Add("Feld 'titel' ist ein Pflichtfeld.");
        }
        else if (titel.Length > 120)
        {
            errors.Add("Feld 'titel' darf maximal 120 Zeichen lang sein.");
        }

        if (!string.IsNullOrEmpty(i.Art) && !Arten.Contains(i.Art))
        {
            errors.Add("Feld 'art' ist keine gueltige Stellenart.");
        }

        if (errors.Count > 0) throw new ValidationException(errors);
    }
}

public sealed record StelleHttpInput(
    string? Titel,
    string? Beschreibung,
    string? Art
);

public sealed record AnlegenResult(int Id, string Titel, string Art, string Status);

File: php/src/StellenangebotService.php
<?php

declare(strict_types=1);

namespace App;

/**
 * Use-Case-Schicht: reine Fachlogik, kennt weder PDO noch HTTP.
 *
 * Wichtigste Geschaeftsregel:
 *   Eine neue Stelle startet IMMER mit Status 'ENTWURF'.
 *   Ein vom Aufrufer gelieferter status wird bewusst ignoriert.
 *   (Die DB hat zwar denselben Default - die Regel hier sichert,
 *    dass auch ein anderer Persistenz-Adapter sie einhaelt.)
 */
final class StellenangebotService
{
    public const STATUS_ENTWURF = 'ENTWURF';

    /** @var list<string> */
    public const ARTEN = [
        'FESTANSTELLUNG',
        'AZUBI',
        'MINIJOB',
        'WERKSTUDENT',
        'PRAKTIKUM',
    ];

    /** @var list<string> */
    public const STATI = [
        'ENTWURF',
        'VEROEFFENTLICHT',
        'GESCHLOSSEN',
        'ARCHIVIERT',
    ];

    public function __construct(
        private readonly StellenangebotRepositoryInterface $repo
    ) {}

    /**
     * Use-Case "Stelle anlegen".
     *
     * @param array<string,mixed> $input
     * @return array{id:int,titel:string,art:string,status:string}
     */
    public function anlegen(array $input): array
    {
        $this->validate($input);

        $titel        = trim((string) $input['titel']);
        $beschreibung = isset($input['beschreibung'])
            ? trim((string) $input['beschreibung'])
            : null;
        $art = isset($input['art']) && $input['art'] !== ''
            ? (string) $input['art']
            : 'FESTANSTELLUNG';

        // Geschaeftsregel: neue Stellen starten IMMER als ENTWURF.
        $status = self::STATUS_ENTWURF;

        $id = $this->repo->insertStelle([
            'titel'        => $titel,
            'beschreibung' => $beschreibung,
            'art'          => $art,
            'status'       => $status,
        ]);

        return [
            'id'     => $id,
            'titel'  => $titel,
            'art'    => $art,
            'status' => $status,
        ];
    }

    /** @return array<int,array<string,mixed>> */
    public function liste(?string $status = null): array
    {
        if ($status !== null && !in_array($status, self::STATI, true)) {
            throw new ValidationException([
                "Parameter 'status' ist kein gueltiger Stellenstatus.",
            ]);
        }
        return $this->repo->listStellen($status);
    }

    /** @param array<string,mixed> $input */
    private function validate(array $input): void
    {
        $errors = [];

        $titel = trim((string) ($input['titel'] ?? ''));
        if ($titel === '') {
            $errors[] = "Feld 'titel' ist ein Pflichtfeld.";
        } elseif (mb_strlen($titel) > 120) {
            $errors[] = "Feld 'titel' darf maximal 120 Zeichen lang sein.";
        }

        if (isset($input['art']) && $input['art'] !== ''
            && !in_array($input['art'], self::ARTEN, true)) {
            $errors[] = "Feld 'art' ist keine gueltige Stellenart.";
        }

        if ($errors !== []) {
            throw new ValidationException($errors);
        }
    }
}
